### DISCUSSION TITLE
typing of a class which returns `callable` from `__new__`

### DISCUSSION THREAD

**Discussion by Brainor** [microsoft/pyright#6299] [Q&A] [answered]
When using Pylance(`2023.10.50`) in VSCode(`1.84.0`) which runs pyright v`1.1.332` with python `3.11.6`, I came across this strange problem regarding the typing.
```python
from collections.abc import Callable
from typing import ParamSpec, TypeVar

_P = ParamSpec("_P")
_R = TypeVar("_R")
def func1(a: int) -> int:
    return a + 1

class TestClass1:
    def __new__(cls, func: Callable[_P, _R]) -> Callable[_P, _R]:
        return func
a1 = TestClass1(func1)  # typing: (variable) a1: TestClass

class TestClass2:
    def __new__(cls, func: Callable[_P, _R]) -> int:
        return 123
a2 = TestClass2(func1)  # typing: (variable) a2: int
```
The typing of `a1` and `a2` is like this:
![image](https://github.com/microsoft/pyright/assets/5477995/82a33732-ed5c-4e01-ac72-e02fabf6b4c8)

`a2` is correctly typed as `int`, but `a1` is typed as `TestClass`, which I want it to be typed as the function `func1`: `(function) def func1(a: int) -> int`. How do I accomplish this? Maybe it's not a pyright thing but a python feature, but any comments are really appreciated.

**Comment by erictraut** [comment, marked as answer]
Pyright's handling of `__new__` includes some heuristics to determine whether it should use the type result of `__new__` or ignore it. That's necessary because `__new__` return results have traditionally been ignored by mypy and other type checkers, so there's a lot of poorly-annotated `__new__` methods in existing source. It's also necessary because some `__new__` return types lead to ambiguities. For example, if the return type of `__new__` is a union that includes an instance of the class and some other class, it's not clear whether the `__init__` call will be invoked at runtime. (The `__init__` is invoked only if `__new__` returns an instance of the same class, so it's ambiguous if `__new__` returns a union.)

The current heuristic rejects (ignores) the `__new__` return type if it's a callable, which is why you're seeing the behavior you're seeing. I think it would be safe to remove this limit because it's unlikely that anyone would unintentionally do what you're trying to do here. After all, it's very odd to return a callable from a `__new__` call. I've never seen it done before.

I'll create an issue out of this discussion thread.

**Reply by Brainor** [reply]
Thank you for your detailed explanation of this particular behavior of `__new__`'s special treatment to the callable (for the fact that instance or other classes are also callable which results to complication). I understand that now, and thank you for the issue #6303 you raised :blush:

------
FYI, the reason I want to do this is because I want to create a class with a `__call__` method acted as a decorator which can be invoked as `@TestClass` and as `@TestClass('hello')` (why it has to be a class is another story). 
With `TestClass` as a decorator to `def func(args):...`, `func(args)` is [semantically equivalent to](https://docs.python.org/3/glossary.html#term-decorator) `TestClass(func)(args)`; the intuitive idea is to override `__new__` method to treat incoming parameter differently, like this:
```python
class TestClass:
    def __new__(cls, message=""):
        if callable(message):
            return TestClass("")(message)
        else:
            return super().__new__(cls)
    def __call__(self, func):
        def inner(*args, **kwds):
            return func(*args, **kwds)
        return inner
@TestClass
def func1():
    ...
@TestClass("hello")
def func2():
    ...
func1()  # typing: (function) func1: TestClass
func2()  # typing: (function) def func2() -> None
```
Even with clear typing of `__new__` or `@typing.overload`, `func1`'s is still not correctly typed, hence this question.
